cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -477,36 +640,59 @@ public boolean apply(RunRecordMeta input) {
     Map<ProgramRunId, RunRecordMeta> programRunIdMap = new LinkedHashMap<>();
     for (Map.Entry<MDSKey, RunRecordMeta> entry : keymap.entrySet()) {
       ProgramId programId = getProgramID(entry.getKey());
-      programRunIdMap.put(new ProgramRunId(programId.getNamespace(), programId.getApplication(),
-                                           programId.getType(), programId.getProgram(), entry.getValue().getPid()),
-                          entry.getValue());
+      programRunIdMap.put(programId.run(entry.getValue().getPid()), entry.getValue());
     }
     return programRunIdMap;
   }
 
-  private  Map<MDSKey, RunRecordMeta> getHistoricalRuns(ProgramId programId, ProgramRunStatus status,
-                                                final long startTime, final long endTime, int limit,
-                                                @Nullable Predicate<RunRecordMeta> filter) {
+  private Map<ProgramRunId, RunRecordMeta> getHistoricalRuns(@Nullable ProgramId programId, ProgramRunStatus status,
+                                                             final long startTime, final long endTime, int limit,
+                                                             @Nullable Predicate<RunRecordMeta> filter) {
     MDSKey historyKey = getProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programId).build();
+    Map<ProgramRunId, RunRecordMeta> programRunIdRunRecordMetaMap = getHistoricalRuns(historyKey, status, startTime,
+                                                                                      endTime, limit, filter);
+
+    int remaining = limit - programRunIdRunRecordMetaMap.size();
+    if (programId != null && programId.getVersion().equals(ApplicationId.DEFAULT_VERSION) && (remaining > 0)) {
+      historyKey = getVersionLessProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programId).build();
+      Map<ProgramRunId, RunRecordMeta> oldRunRecordMetaMap = getHistoricalRuns(historyKey, status, startTime, endTime,
```

Comment:
Instead of doing two scans all the time, we can change the scan we do when programId is default version. In that case do only the second scan, and add a filter that returns only the right records.